File: backend_samples/backend/react/app/_lib/app_insights_server.ts
/**
 * Server-side Application Insights wrapper.
 *
 * Uses the Node SDK ('applicationinsights') — the existing `components/appinsights.tsx`
 * is the browser SDK and does not emit telemetry from server code (API routes,
 * server components).
 *
 * Configure via APPLICATIONINSIGHTS_CONNECTION_STRING (preferred). For backwards
 * compatibility the NEXT_PUBLIC_APPINSIGHTS_INSTRUMENTATION_KEY env var is also
 * accepted. The attestation service uses a console logger when neither is set.
 */

import * as appInsights from 'applicationinsights';
import type { TelemetryClient } from 'applicationinsights';

let initAttempted = false;
let client: TelemetryClient | null = null;

function initIfNeeded(): TelemetryClient | null {
  if (initAttempted) return client;
  initAttempted = true;

  const connectionString =
    process.env.APPLICATIONINSIGHTS_CONNECTION_STRING ||
    (process.env.NEXT_PUBLIC_APPINSIGHTS_INSTRUMENTATION_KEY
      ? `InstrumentationKey=${process.env.NEXT_PUBLIC_APPINSIGHTS_INSTRUMENTATION_KEY}`
      : '');

  if (!connectionString) {
    console.warn(
      'App Insights server SDK not configured: APPLICATIONINSIGHTS_CONNECTION_STRING and NEXT_PUBLIC_APPINSIGHTS_INSTRUMENTATION_KEY are both unset.'
    );
    return null;
  }

  try {
    appInsights
      .setup(connectionString)
      .setAutoCollectConsole(true, true)
      .setAutoCollectExceptions(true)
      .setAutoCollectRequests(true)
      .setAutoCollectDependencies(true)
      .setAutoCollectPerformance(true, true)
      .setSendLiveMetrics(false)
      .start();
    client = appInsights.defaultClient;
    if (client) {
      client.context.tags[client.context.keys.cloudRole] =
        process.env.WEBSITE_SITE_NAME || 'liveness-webapp';
    }
  } catch (err) {
    console.error('App Insights init failed:', err);
    client = null;
  }
  return client;
}

type Props = Record<string, unknown> | undefined;

function stringifyProps(props: Props): Record<string, string> | undefined {
  if (!props) return undefined;
  const out: Record<string, string> = {};
  for (const [k, v] of Object.entries(props)) {
    if (v === undefined || v === null) continue;
    if (typeof v === 'string') {
      out[k] = v;
    } else if (typeof v === 'number' || typeof v === 'boolean') {
      out[k] = String(v);
    } else {
      try {
        out[k] = JSON.stringify(v);
      } catch {
        out[k] = String(v);
      }
    }
  }
  return out;
}

function numericMeasurements(
  measurements?: Record<string, number>
): Record<string, number> | undefined {
  if (!measurements) return undefined;
  const out: Record<string, number> = {};
  for (const [k, v] of Object.entries(measurements)) {
    if (typeof v === 'number' && Number.isFinite(v)) out[k] = v;
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

export function trackEvent(
  name: string,
  properties?: Props,
  measurements?: Record<string, number>
): void {
  const c = initIfNeeded();
  if (!c) return;
  try {
    c.trackEvent({
      name,
      properties: stringifyProps(properties),
      measurements: numericMeasurements(measurements),
    });
  } catch (err) {
    console.error('trackEvent failed:', err);
  }
}

export function trackException(error: unknown, properties?: Props): void {
  const c = initIfNeeded();
  if (!c) return;
  try {
    const exception =
      error instanceof Error
        ? error
        : new Error(typeof error === 'string' ? error : JSON.stringify(error));
    c.trackException({
      exception,
      properties: stringifyProps(properties),
    });
  } catch (err) {
    console.error('trackException failed:', err);
  }
}

export function trackMetric(name: string, value: number, properties?: Props): void {
  const c = initIfNeeded();
  if (!c) return;
  if (!Number.isFinite(value)) return;
  try {
    c.trackMetric({ name, value, properties: stringifyProps(properties) });
  } catch (err) {
    console.error('trackMetric failed:', err);
  }
}

export function trackRequest(opts: {
  name: string;
  url: string;
  duration: number;
  resultCode: string | number;
  success: boolean;
  source?: string;
  properties?: Props;
}): void {
  const c = initIfNeeded();
  if (!c) return;
  try {
    const properties =
      opts.source !== undefined
        ? { ...(opts.properties ?? {}), source: opts.source }
        : opts.properties;
    c.trackRequest({
      name: opts.name,
      url: opts.url,
      duration: opts.duration,
      resultCode: String(opts.resultCode),
      success: opts.success,
      properties: stringifyProps(properties),
    });
  } catch (err) {
    console.error('trackRequest failed:', err);
  }
}

export function trackDependency(opts: {
  name: string;
  target?: string;
  data?: string;
  duration: number;
  success: boolean;
  resultCode?: string | number;
  dependencyTypeName?: string;
  properties?: Props;
}): void {
  const c = initIfNeeded();
  if (!c) return;
  try {
    c.trackDependency({
      name: opts.name,
      target: opts.target || '',
      data: opts.data || '',
      duration: opts.duration,
      success: opts.success,
      resultCode: String(opts.resultCode ?? (opts.success ? '0' : '1')),
      dependencyTypeName: opts.dependencyTypeName || 'HTTP',
      properties: stringifyProps(opts.properties),
    });
  } catch (err) {
    console.error('trackDependency failed:', err);
  }
}

/**
 * Flush any queued telemetry. Call before short-lived processes exit so events
 * aren't dropped. Safe to call when not configured (no-op).
 */
export async function flush(): Promise<void> {
  const c = initIfNeeded();
  if (!c) return;
  try {
    await c.flush();
  } catch (err) {
    console.error('App Insights flush failed:', err);
  }
}

File: backend_samples/backend/python/app/telemetry/app_insights_server.py
"""
Server-side Application Insights wrapper.

Wraps the Azure Monitor OpenTelemetry distro when
APPLICATIONINSIGHTS_CONNECTION_STRING is set; otherwise helpers write to the
application's console logger. The interface (trackEvent / trackException /
trackDependency / trackRequest / flush) is kept name-compatible with the Node
module so the rest of the port reads the same.
"""

from __future__ import annotations

import json
import logging
import os
import time
import traceback
from typing import Any, Mapping, Optional

log = logging.getLogger("ai")

_init_attempted = False
_logger: Optional[logging.Logger] = None
_meter = None


def _init_if_needed():
    global _init_attempted, _logger
    if _init_attempted:
        return _logger
    _init_attempted = True

    conn = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING", "")
    if not conn:
        log.warning(
            "App Insights server SDK not configured: "
            "APPLICATIONINSIGHTS_CONNECTION_STRING is unset."
        )
        return None

    try:
        from azure.monitor.opentelemetry import configure_azure_monitor

        configure_azure_monitor(connection_string=conn)
        _logger = logging.getLogger("appinsights")
        return _logger
    except Exception as err:
        log.error("App Insights init failed: %s", err)
        _logger = None
        return None


def _stringify_props(props: Optional[Mapping[str, Any]]) -> dict[str, str]:
    if not props:
        return {}
    out: dict[str, str] = {}
    for k, v in props.items():
        if v is None:
            continue
        if isinstance(v, (str, int, float, bool)):
            out[k] = str(v)
        else:
            try:
                out[k] = json.dumps(v, default=str)
            except Exception:
                out[k] = str(v)
    return out


def track_event(
    name: str,
    properties: Optional[Mapping[str, Any]] = None,
    measurements: Optional[Mapping[str, float]] = None,
) -> None:
    c = _init_if_needed()
    payload = {"event": name, "properties": _stringify_props(properties)}
    if measurements:
        payload["measurements"] = {
            k: v for k, v in measurements.items() if isinstance(v, (int, float))
        }
    if c:
        c.info("Event %s", json.dumps(payload, default=str))
    else:
        log.info("[Attestation] Event %s", json.dumps(payload, default=str))


def track_exception(error: Any, properties: Optional[Mapping[str, Any]] = None) -> None:
    c = _init_if_needed()
    payload = {
        "exception": str(error),
        "stack": "".join(
            traceback.format_exception(type(error), error, error.__traceback__)
        )
        if isinstance(error, BaseException)
        else None,
        "properties": _stringify_props(properties),
    }
    if c:
        c.error("Exception %s", json.dumps(payload, default=str))
    else:
        log.error("[Attestation] Exception %s", json.dumps(payload, default=str))


def track_metric(name: str, value: float, properties: Optional[Mapping[str, Any]] = None) -> None:
    c = _init_if_needed()
    payload = {"name": name, "value": value, "properties": _stringify_props(properties)}
    if c:
        c.info("Metric %s", json.dumps(payload))
    else:
        log.info("[Attestation] Metric %s", json.dumps(payload))


def track_request(
    name: str,
    url: str,
    duration: float,
    result_code: Any,
    success: bool,
    source: Optional[str] = None,
    properties: Optional[Mapping[str, Any]] = None,
) -> None:
    c = _init_if_needed()
    props = dict(properties or {})
    if source is not None:
        props["source"] = source
    payload = {
        "request": name,
        "url": url,
        "durationMs": duration,
        "resultCode": str(result_code),
        "success": success,
        "properties": _stringify_props(props),
    }
    if c:
        c.info("Request %s", json.dumps(payload, default=str))
    else:
        log.info("[Attestation] Request %s", json.dumps(payload, default=str))


def track_dependency(
    name: str,
    target: str = "",
    data: str = "",
    duration: float = 0,
    success: bool = True,
    result_code: Any = None,
    dependency_type_name: str = "HTTP",
    properties: Optional[Mapping[str, Any]] = None,
) -> None:
    c = _init_if_needed()
    payload = {
        "dependency": name,
        "target": target,
        "data": data,
        "durationMs": duration,
        "success": success,
        "resultCode": str(result_code if result_code is not None else ("0" if success else "1")),
        "type": dependency_type_name,
        "properties": _stringify_props(properties),
    }
    if c:
        c.info("Dependency %s", json.dumps(payload, default=str))
    else:
        log.info("[Attestation] Dependency %s", json.dumps(payload, default=str))


def now_ms() -> int:
    return int(time.time() * 1000)


async def flush() -> None:
    pass
